engine/engine.go

```diff
@@ -95,23 +92,42 @@ func NewFromSettings(settings *Settings) (*Engine, error) {
 		gctlog.Infoln(gctlog.Global, "Logger initialised.")
 	}
 
-	b.Settings.ConfigFile = filePath
-	b.Settings.DataDir = settings.DataDir
+	b.Settings.ConfigFile = settings.ConfigFile
+	b.Settings.DataDir = b.Config.GetDataPath()
 	b.Settings.CheckParamInteraction = settings.CheckParamInteraction
 
 	err = utils.AdjustGoMaxProcs(settings.GoMaxProcs)
 	if err != nil {
 		return nil, fmt.Errorf("unable to adjust runtime GOMAXPROCS value. Err: %s", err)
 	}
 
-	ValidateSettings(&b, settings)
+	validateSettings(&b, settings)
 	return &b, nil
 }
 
-// ValidateSettings validates and sets all bot settings
-func ValidateSettings(b *Engine, s *Settings) {
-	flag.Visit(func(f *flag.Flag) { flagSet[f.Name] = true })
+// loadConfigWithSettings creates configuration based on the provided settings
+func loadConfigWithSettings(settings *Settings) (*config.Config, error) {
+	filePath, err := config.GetFilePath(settings.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Loading config file %s..\n", filePath)
+
+	conf := &config.Cfg
+	err = conf.ReadConfig(filePath, settings.EnableDryRun)
+	if err != nil {
+		return nil, fmt.Errorf(config.ErrFailureOpeningConfig, filePath, err)
+	}
+	// Apply overrides from settings
+	if flagSet["datadir"] {
+		conf.DataDir = settings.DataDir
+	}
```

Comment:
I don't think we should be overwriting our configuration file. This needs a bit of discussion.

Reply:
I like the following logic:
a) If the datadir flag isn't set and the config contains no datadir value, use the default data directory and update the config to use it.
b) If the datadir flag is set, skip the config setting and use the flag value, but don't save this updated datadir in the config. The user should update the datadir setting in their config if they want a persistent datadir location

Reply:
a) is already the case
b) yep, this was my initial idea as well, but... 
* without this change the -datadir isn't properly applied at all - (logs and scripts are using default regardless of the flag), so for that at least the current change is an improvement I think
* the initialisation of the directories is done in the 'config' package without any dependency on engine and Settings, this would have to be turned around and the `(c *Config) Check*()` functions would have to be in the engine instead. What is your policy on breaking changes? As this would change a lot.

Reply:
For b) You could add `settings.EnableDryRun = true` inside the above if statement and a log warn that dryrun is enabled for when its applied. We do have `func dryrunParamInteraction("datadir")`, but Bot is not setup yet at this time to use it. Using  `settings.EnableDryRun = true` will preserve the user's settings while allowing you to have a custom datadir for that run, which properly creates directories, logs, script outputs etc.

Reply:
Done.